### PR TITLE
Remote Keyboard added

### PR DESCRIPTION
**Pull request by derFogel**
I added Remote input with a virtual keyboard with different layouts.
Currently there are only two layouts, but it is easy to add new ones (simply place a json file in keyboard-layouts/).

I tried to copy your coding style but I am not sure if i matched your expectations.

### CODE REVIEW

**Comment by R1tschY** [comment]
Thank you for your pull request.

Why do you not use the system keyboard? For example through an invisible text control. This would remove the burden for creating and maintaining keyboard layouts and all the logic behind it. IMHO it is not necessary to create an own keyboard implementation. When it is possible, please use the system keyboard.

**Comment by derFogel** [comment]
I made a own keyboard because the default keyboard is missing some keys (like ctrl, alt, arrows).
But yes, it is easier to use the default keyboard.
I will remove my own keyboard and use the system keyboard. Maybe it is possible to add these missing keys as an extra row to the default keyboard. Since it is based on the maliit keyboard, there should be something like a plugin api. I will talk with on of the developers of the maliit keyboard and see if something like this is possible.

**Comment by R1tschY** [comment]
I did not thought about the extra keys. Maybe a hybrid approach makes sense: A keyboard with ASCII chars and control keys (ctrl, alt, win, arrow keys, ...) like the Terminal app keyboard (for shortcuts?) and the possibility to open the system keyboard to write texts in the user language.

File: app/src/ui/keyboardlayoutprovider.cpp [hunk 1]
```diff
@@ -117,3 +119,55 @@ QVariantList KeyboardLayoutProvider::row6() const
...
+}
+
+void KeyboardLayoutProvider::releaseFeedback()
+{
+    QFeedbackHapticsEffect::playThemeEffect(QFeedbackEffect::ReleaseWeak);
+}
+
+void KeyboardLayoutProvider::saveConfig() const
+{
+    QFile config(QStandardPaths::writableLocation(QStandardPaths::ConfigLocation) + "/harbour-sailfishconnect/keyboardlayout.conf");
```
Comment (R1tschY) [inline app/src/ui/keyboardlayoutprovider.cpp]: Please use QSettings for application-wide settings. Use "RemoteKeyboard" as group. As example see the runInBackground setting in the UI class: 
https://github.com/R1tschY/harbour-sailfishconnect/blob/2d6145c5f084370251d66b5b6a6e8f6cf1223528/app/src/ui.cpp#L49
https://github.com/R1tschY/harbour-sailfishconnect/blob/2d6145c5f084370251d66b5b6a6e8f6cf1223528/app/src/ui.cpp#L138

**Comment by derFogel** [comment]
I modified the keyboard layout provider to use the default layout from jolla (they are stored in /usr/share/maliit/plugins/com/jolla/layouts/). It automatically adds the control and alt key next to the spacebar and adds the arrow keys when switching to the character page. I think this is a good solution because we don't have to maintain different keyboard layouts and every user can use their favourite layout.
I also changed my settings implementation (forgot about QSettings) to QSettings.

**Comment by R1tschY** [comment]
Is your pull request ready for review?

**Comment by derFogel** [comment]
yes

**Comment by R1tschY** [comment]
@derFogel: Could you please resolve the conflicts in app.pro.

File: app/src/ui/keyboardlayoutprovider.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+{
+    return m_layout;
+}
+
+void KeyboardLayoutProvider::setLayout(const QString &layout)
+{
+    // seems to be the only way to get the files stored
+    QFile layoutFile("/usr/share/maliit/plugins/com/jolla/layouts/" + layout + ".qml");
+    if (!layoutFile.open(QIODevice::ReadOnly)) {
+        qDebug() << "unkown layout: " << layout;
```
Comment (R1tschY) [inline app/src/ui/keyboardlayoutprovider.cpp]: Typo and please use an own logger for KeyboardLayoutProvider (see other files in app/src/ui).
```suggestion
        qCDebug(logger) << "Unknown layout:" << layout;
```

File: app/qml/components/KeyboardPage.qml [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                        markValue: modelData["caption"] === "ctrl" ? modifiers.ctrl : (modelData["caption"] === "alt" && !page.chars ? modifiers.alt : 0)
+
+                        Image {
+                            anchors.centerIn: parent
+                            width: parent.width
+                            height: parent.height
+                            fillMode: Image.PreserveAspectFit
+                            visible: (modelData["symView"] === "left" || modelData["symView"] === "down" || modelData["symView"] === "right") && page.chars
+
+                            source:  "image://theme/icon-m-" + modelData["symView"]
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: I'm getting:
`[W] unknown:197 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardPage.qml:197:25: QML Image: Failed to get image from provider: image://theme/icon-m-undefined`

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    stepSize: 50
+                    value: _keyboardLayout.repeatInterval
+                    valueText: qsTr("%1ms repeat interval").arg(value)
+
+                    onValueChanged: {
+                        _keyboardLayout.repeatInterval = value;
+                    }
+                }
+
+                TextSwitch {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml]: [W] unknown:74 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardChangeLayoutPage.qml:74:17: QML TextSwitch: Binding loop detected for property "checked"

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                }
+
+                onClicked: {
+                    _keyboardLayout.layout = modelData;
+                }
+            }
+
+            footer: Column {
+
+                Slider {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml]: [W] unknown:61 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardChangeLayoutPage.qml:61:17: QML Slider: Binding loop detected for property "value"

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+
+Page {
+    allowedOrientations: Orientation.All
+
+    SilicaFlickable {
+        anchors.fill: parent
+
+        ListView {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml]: Please use SilicaListView
```suggestion
        SilicaListView {
```

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+
+Page {
+    allowedOrientations: Orientation.All
+
+    SilicaFlickable {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml:24]: This SilicaFlickable is unneeded. Place everything into the SilicaListView.

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                    color: modelData == _keyboardLayout.layout ? Theme.highlightColor : Theme.secondaryColor
+                    font.pixelSize: Theme.fontSizeMedium
+                }
+
+                onClicked: {
+                    _keyboardLayout.layout = modelData;
+                }
+            }
+
+            footer: Column {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml]: Please put it in the header. So it is a little bit hidden at the end of a long list.

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+
+Page {
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml:21]: I think putting the layout list into a ComboBox provides a better workflow. What do you think?

File: app/qml/components/KeyboardPage.qml [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+                                plugin.sendKeyPress("right", false, modifiers.ctrl, modifiers.alt);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // I hate this hack but it seems this is the only way to push an attachted page
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: I would prefer the use of the pulley menu.

File: app/src/plugins/remoteinput/remoteinput.cpp [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    //{Qt::Key_F6, 26},
+    //{Qt::Key_F7, 27},
+    //{Qt::Key_F8, 28},
+    //{Qt::Key_F9, 29},
+    //{Qt::Key_F10, 30},
+    //{Qt::Key_F11, 31},
+    //{Qt::Key_F12, 32},
+};
+
+SailfishConnect::RemoteInputPlugin::RemoteInputPlugin(Device *device, const QString &name, const QSet<QString> &outgoingCapabilities) : KdeConnectPlugin(device, name, outgoingCapabilities)
```
Comment (R1tschY) [inline app/src/plugins/remoteinput/remoteinput.cpp]: Please do not create lines with more than 80 charaters in C++ files. You can configure this limit in QtCreator. Please look through all your code to fix this.

File: app/src/plugins/remoteinput/remoteinput.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
+#include "remoteinput.h"
```
Comment (R1tschY) [inline app/src/plugins/remoteinput/remoteinput.cpp]: Missing copyright header. Please look through all your code to fix this.

**Review by R1tschY — CHANGES_REQUESTED** [review]
I did a short review of your changes.
Generally: Feedback and Emoij layout did not work for me.

File: app/qml/components/KeyboardChangeLayoutPage.qml (hunk 9 above)
Comment (derFogel) [reply app/qml/components/KeyboardChangeLayoutPage.qml:21]: It is definitely better. Thanks for the suggestion

**Comment by derFogel** [comment]
@R1tschY done

File: app/qml/components/KeyboardPage.qml [hunk 13]
```diff
@@ -53,6 +53,15 @@ Page {
     SilicaFlickable {
         anchors.fill: parent
 
+        PullDownMenu {
+            MenuItem {
+                text: qsTr("Change layout")
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Rename it to "Settings" or make two entries: "Change layout" and "Settings"

File: app/src/plugins/remoteinput/remoteinput.cpp [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    //{Qt::Key_F12, 32},
+};
+
+SailfishConnect::RemoteInputPlugin::RemoteInputPlugin(Device *device,
+        const QString &name, const QSet<QString> &outgoingCapabilities)
+    : KdeConnectPlugin(device, name, outgoingCapabilities)
+{
+}
+
+bool SailfishConnect::RemoteInputPlugin::receivePacket(const NetworkPacket &np)
```
Comment (R1tschY) [inline app/src/plugins/remoteinput/remoteinput.cpp]: Please handle "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate" like in:
https://github.com/KDE/kdeconnect-kde/blob/7df58d0d04b5e7a1f99c3d35a8fd485bc83e868c/plugins/remotekeyboard/remotekeyboardplugin.cpp#L79

File: app/src/plugins/remoteinput/metadata.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
+    "Id": "SailfishConnect::RemoteInputPlugin",
```
Comment (R1tschY) [inline app/src/plugins/remoteinput/metadata.json]: Please, rename every thing to remotekeyboard / remotekeyboardplugin / RemoteKeyboardPlugin. So the names match with KDE-Connect.

File: app/qml/components/KeyboardPage.qml [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+        property int shift: 0
+        property int ctrl: 0
+        property int alt: 0
+    }
+    property bool chars: false
+
+    function displayCorrectChar(data) {
+        if (page.chars && typeof data["symView"] !== "undefined") {
+            // dont show text on arrow keys
+            if (data["symView"] === "up" || data["symView"] === "left" || data["symView"] === "down" || data["symView"] === "right") return " ";
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Please use something like ['a', 'b', 'c'].indexOf(str) >= 0 or str.match("a|b|c") to reduce line length

File: app/qml/components/KeyboardPage.qml [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+                        }
+
+                        onClicked: {
+                            if (label == "?123" || label == "ABC") {
+                                page.chars = !page.chars;
+                                modifiers.shift = 0;
+                                modifiers.alt = 0;
+                                modifiers.ctrl = 0;
+                            } else if (label == "alt") {
+                                modifiers.alt++;
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Just `modifiers.alt = (modifiers.alt + 1) % 3`

File: app/qml/components/KeyboardPage.qml [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+                            height: parent.height
+                            fillMode: Image.PreserveAspectFit
+                            visible: (modelData["caption"] === "shift" && !page.chars) || modelData["caption"] === "backspace" || (modelData["symView"] === "up" && page.chars)
+
+                            source: modelData["symView"] === "up" ? "image://theme/icon-m-up" : (modelData["caption"] === "shift" ? "image://theme/icon-m-capslock" : "image://theme/icon-m-backspace")
+                        }
+
+                        onClicked: {
+                            if (modelData["caption"] === "shift") {
+                                modifiers.shift++;
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Just `modifiers.shift = (modifiers.shift + 1) % 3`

File: app/qml/components/KeyboardPage.qml [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+                                return displayCorrectChar(data);
+                            }
+                        }
+
+                        Image {
+                            anchors.centerIn: parent
+                            width: parent.width
+                            height: parent.height
+                            fillMode: Image.PreserveAspectFit
+                            visible: (modelData["caption"] === "shift" && !page.chars) || modelData["caption"] === "backspace" || (modelData["symView"] === "up" && page.chars)
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Please break lines in qml files for more readability (<= 100 good, <= 80 better).

File: app/src/ui/keyboardlayoutprovider.cpp [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+    return m_layout;
+}
+
+void KeyboardLayoutProvider::setLayout(const QString &layout)
+{
+    // seems to be the only way to get the files stored
+    QFile layoutFile("/usr/share/maliit/plugins/com/jolla/layouts/" + layout
+                     + ".qml");
+    if (!layoutFile.open(QIODevice::ReadOnly)) {
+        qCDebug(logger) << "unkown layout: " << layout;
```
Comment (R1tschY) [inline app/src/ui/keyboardlayoutprovider.cpp]:
```suggestion
        qCDebug(logger) << "Unknown layout: " << layout;
```

File: app/src/ui.cpp [hunk 21]
```diff
@@ -79,6 +80,8 @@ void UI::showMainWindow()
     // view
     m_view->rootContext()->setContextProperty("daemon", m_daemon);
     m_view->rootContext()->setContextProperty("ui", this);
+    m_view->rootContext()->setContextProperty("_keyboardLayout", m_keyboardLayoutProvider);
```
Comment (R1tschY) [inline app/src/ui.cpp]: Why use an underscore? 
```suggestion
    m_view->rootContext()->setContextProperty("keyboardLayout", m_keyboardLayoutProvider);
```

File: app/qml/components/KeyboardPage.qml [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+                                page.chars = !page.chars;
+                                modifiers.shift = 0;
+                                modifiers.alt = 0;
+                                modifiers.ctrl = 0;
+                            } else if (label == "alt") {
+                                modifiers.alt++;
+                                if (modifiers.alt > 2) modifiers.alt = 0;
+                            } else if (label == "ctrl") {
+                                modifiers.ctrl++;
+                                if (modifiers.ctrl > 2) modifiers.ctrl = 0;
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: `modifiers.alt = (modifiers.alt + 1) % 3`

File: app/qml/components/KeyboardPage.qml [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+
+                        onClicked: {
+                            if (label == "?123" || label == "ABC") {
+                                page.chars = !page.chars;
+                                modifiers.shift = 0;
+                                modifiers.alt = 0;
+                                modifiers.ctrl = 0;
+                            } else if (label == "alt") {
+                                modifiers.alt++;
+                                if (modifiers.alt > 2) modifiers.alt = 0;
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: `modifiers.alt = (modifiers.alt + 1) % 3`

File: app/src/plugins/remoteinput/remoteinput.cpp [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+{
+    Q_UNUSED(np);
+    return false;
+}
+
+void SailfishConnect::RemoteInputPlugin::sendKeyPress(const QString &key,
+    bool shift, bool ctrl, bool alt) const
+{
+    NetworkPacket np("kdeconnect.mousepad.request", {{"key", key},
+                        {"specialKey", specialKeysMap[key]}, {"shift", shift},
```
Comment (R1tschY) [inline app/src/plugins/remoteinput/remoteinput.cpp]: Avoid creating new keys in specialKeysMap:
```suggestion
                        {"specialKey", specialKeysMap.value(key, 0)}, {"shift", shift},
```

File: app/src/ui.cpp [hunk 25]
```diff
@@ -79,6 +80,8 @@ void UI::showMainWindow()
...
+    qmlRegisterType<KeyboardLayoutProvider>("com.meego.maliitquick", 1, 0, "Layout");
```
Comment (R1tschY) [inline app/src/ui.cpp]: Please do not reuse others namespaces. Please move the code to registerQmlTypes() in sailfishconnect.cpp.
```suggestion
    qmlRegisterType<KeyboardLayoutProvider>("SailfishConnect.RemoteKeyboard", 0, 4, "KeyboardLayout");
```

File: app/src/ui/keyboardlayoutprovider.cpp [hunk 26]
```diff
@@ -0,0 +1,240 @@
+/*
```
Comment (R1tschY) [inline app/src/ui/keyboardlayoutprovider.cpp]: Please move it to the plugins/remotekeyboard dir.

File: app/app.pro [hunk 27]
```diff
@@ -80,7 +82,12 @@ DISTFILES += \
     qml/components/ShareUi.qml \
     qml/pages/TransfersPage.qml \
     src/plugins/telephony/metadata.json \
-    src/dbus/ofono/org.ofono.Manager.xml
+    src/dbus/ofono/org.ofono.Manager.xml \
+    qml/components/KeyboardPage.qml \
+    qml/components/Key.qml \
+    qml/components/RemoteInput.qml \
+    src/plugins/remoteinput/metadata.json \
+    keyboard-layouts/english.json
```
Comment (R1tschY) [inline app/app.pro]: This file does not exist.

**Review by R1tschY — CHANGES_REQUESTED** [review]
I played with it and missed the ESC key. Also there are many other keys missing (https://github.com/KDE/kdeconnect-kde/blob/7df58d0d04b5e7a1f99c3d35a8fd485bc83e868c/plugins/remotekeyboard/remotekeyboardplugin.cpp#L32). Any possibility to add a new row at the top?

At the moment landscape orientation is forced. I think it is a little bit strict. Why not allow other orientations?

Keep going. Looks promising. :)

**Comment by derFogel** [comment]
I am not sure how to handle the "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate". The official andoid app also does not handled them (see https://github.com/KDE/kdeconnect-android/blob/master/src/org/kde/kdeconnect/Plugins/MousePadPlugin/KeyListenerView.java#L99).

**Comment by R1tschY** [comment]
> I am not sure how to handle the "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate". The official andoid app also does not handled them (see https://github.com/KDE/kdeconnect-android/blob/master/src/org/kde/kdeconnect/Plugins/MousePadPlugin/KeyListenerView.java#L99).

Just return true, so there is no warning about a not processed message.

**Comment by derFogel** [comment]
ok, done

**Comment by R1tschY** [comment]
Next week I will do a final review.

File: app/src/plugins/remotekeyboard/keyboardlayoutprovider.cpp [hunk 28]
```diff
@@ -0,0 +1,259 @@
...
+    m_row5 = row5.toVariantList();
+
+    m_layout = layout;
+
+    emit layoutChanged();
+}
+
+QVariantList KeyboardLayoutProvider::layouts()
+{
+    QDir layoutDir("/usr/share/maliit/plugins/com/jolla/layouts/");
```
Comment (R1tschY) [inline app/src/plugins/remotekeyboard/keyboardlayoutprovider.cpp]: When parsing "/usr/share/maliit/plugins/com/jolla/layouts*.conf" (QSettings can be used for parsing) you get all layouts with the correct language name.

File: app/qml/components/KeyboardChangeLayoutPage.qml [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+    allowedOrientations: Orientation.All
+
+    SilicaFlickable {
+        anchors.fill: parent
+
+        Column {
+            anchors.fill: parent
+
+            PageHeader {
+                title: qsTr("Choose layout")
```
Comment (R1tschY) [inline app/qml/components/KeyboardChangeLayoutPage.qml]:
```suggestion
                title: qsTr("Settings")
```

File: app/qml/components/KeyboardPage.qml [hunk 30]
```diff
@@ -0,0 +1,278 @@
...
+            return data["symView"];
+        } else {
+            if (modifiers.shift && typeof data["captionShifted"] !== "undefined") {
+                return data["captionShifted"];
+            }
+            return data["caption"];
+        }
+    }
+
+    PageHeader {
```
Comment (R1tschY) [inline app/qml/components/KeyboardPage.qml]: Please use as header of SilicaFlickable:
```
SilicaFlickable {
  header: PageHeader { ...
```

**Review by R1tschY — CHANGES_REQUESTED** [review]
I found out that the language layout selection page is blank, when opening in landscape. But is not blank, when changing to landscape when page is already open. Would you fix this?

Using the layouts_*.conf files is not important. You can do it if you want.

File: app/qml/components/KeyboardPage.qml (hunk 30 above)
Comment (derFogel) [reply app/qml/components/KeyboardPage.qml]: Does not work: Error: Error while loading page: file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardPage.qml:52 Cannot assign to non-existent property "header"

**Comment by derFogel** [comment]
Done, now the displayed names are loaded from the config files

**Review by R1tschY — APPROVED** [review]

**Pull request merged**